handle try catch in different scopes, closes #464

Pangoraw opened the pull request:
Hi,

I added a case to handle `try`/`catch` expressions as two blocks with their own scopes. This prevents the error happening in #464.

We start by exploring the `catch` block to deal with the case where a global reference could be shadowed with the symbol for the catched error:
```julia
a = 1
```
```julia
try
   do_something()
catch a
   report(a)
end
```

We then explore the `try` block as usual.

fonsp commented:
Great, thanks again! Can you add tests with `finally`? And one with `finally` but without `catch`?

Pangoraw commented:
I had not accounted for `finally` :facepalm: It should be all set now.

> That's allowed, right?

It look like [it is](https://docs.julialang.org/en/v1/manual/control-flow/#finally-Clauses), I added tests for this cases. 

Since I had a bit of code repetition, I added a helper function to explore the code in a new scope.

fonsp commented on src/analysis/ExpressionExplorer.jl:
```diff
@@ -112,6 +112,14 @@ Base.push!(x::Set) = x
 # HELPER FUNCTIONS
 ###
 
+function explore_new_scope(ex::Expr, scopestate::ScopeState)::SymbolsState
```
rename to explore_inner_scoped

fonsp commented on src/analysis/ExpressionExplorer.jl:
```diff
@@ -112,6 +112,14 @@ Base.push!(x::Set) = x
+    innerscopestate = deepcopy(scopestate)
+    innerscopestate.inglobalscope = false
+
+    return explore!(Expr(:block, ex), innerscopestate)
```
why create a `:block`?

fonsp commented on src/analysis/ExpressionExplorer.jl:
```diff
@@ -112,6 +112,14 @@ Base.push!(x::Set) = x
+    innerscopestate.inglobalscope = false
+
+    return explore!(Expr(:block, ex), innerscopestate)
+end
```
This pattern is used in lots of other places. Maybe you could also replace those with a call to `explore_inner_scoped`?

Pangoraw replied:
I replaced it for `let`, `for` and `while` :+1: However, I could not use it on the [function declaration](https://github.com/fonsp/Pluto.jl/blob/master/src/analysis/ExpressionExplorer.jl#L386-L391) where it seemed different enough.

fonsp commented:
Awesome, thanks!!
